### ISSUE TITLE
502 Bad Gateway after latest webapp update

### ISSUE THREAD

**Issue by anders0ndg** [nutechsoftware/alarmdecoder-webapp#60]
The latest webapp update causes an issue with the webapp which results in a "502 Bad Gateway" message in the browser. I updated what had been a stable version, and then attempted to troubleshoot to no avail. I downloaded the current image (alarm decoder-raspbian-2018-05-29.img) and started over. After going through the initial config I updated the webapp again through alarmdecoder.local/update/ with the same result after restart (IE: 502 Bad Gateway). Unfortunately my ability to troubleshoot further is limited.

**Comment by dcmeglio** [comment]
I'm having the exact same issue. Worried I bricked my device

**Comment by f34rdotcom** [comment]
Never had one get bricked so don't worry about that. Do you have backups of your settings from the AD2 webapp?

This error is usually easy to fix by repairing the /opt/alarmdecoder and /opt/alarmdecoder-webapp/ 

I can help if needed remotely just call or email support I am available most days.

**Comment by anders0ndg** [comment]
Thanks. I re-imaged the SD card 3 times using the latest image for the AD2PI. The 3rd time I just updated the alarmdecoder libraries and have not applied the latest alarmdecoder-webapp libraries. All is good so far, other than still working through my SmartThings integration again. However, I am sure that if I apply the webapp library updates it will “break” the webapp. Any help there would be welcome as it shows that it’s 37 commits behind.

**Comment by f34rdotcom** [comment]
> Thanks. I re-imaged the SD card 3 times using the latest image for the AD2PI. The 3rd time I just updated the alarmdecoder libraries and have not applied the latest alarmdecoder-webapp libraries. All is good so far, other than still working through my SmartThings integration again. However, I am sure that if I apply the webapp library updates it will “break” the webapp. Any help there would be welcome as it shows that it’s 37 commits behind.
> […](#)
> On May 29, 2019, at 12:03 AM, Sean Mathews ***@***.***> wrote: Never had one get bricked so don't worry about that. Do you have backups of your settings from the AD2 webapp? This error is usually easy to fix by repairing the /opt/alarmdecoder and /opt/alarmdecoder-webapp/ I can help if needed remotely just call or email support I am available most days. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or mute the thread.

Hmm.. Ok I will try and find time today to setup a test again. Last I tried from a fresh image the update worked :(  If possible could I get your help since you have a setup ready to test to try and upgrade it while remote access your system and look at the logs? 503 443 2000 x 703 ad2usb@support.nutech.com I "should" be here most of the day and I have been at the office late most nights working on projects. Just give me a call.

Best
 Sean

**Comment by anders0ndg** [comment]
Sure. I typically do not get home until 6:30 to 7:00 PM CDT. Would you be available after 7:00 this evening?

Thanks,
Duwayn
918-671-5322

**Comment by f34rdotcom** [comment]
Yes I will stay till 9PM PST tonight. Talk to you then.
Best
 SM

**Comment by anders0ndg** [comment]
Thank you!

**Comment by neydah700** [comment]
I am having the same issue. Just updated WebApp. Now 502 Bad Gateway error.

**Comment by dcmeglio** [comment]
@f34rdotcom as far as repairing goes, I followed the steps listed at https://github.com/nutechsoftware/alarmdecoder-webapp/issues/48 which I think are what you meant and it still did not work. Same 502 error

**Comment by bradpeczka** [comment]
I had the same issue on Friday last week. I reimaged my Pi with the latest Stretch build (PIBAKERY-AD2PIBOOT-RASPBIAN-9-STRETCH-20181015) since it was due anyway, and performed the library and webapp updates post-install. No issues thus far, everything works as it should.

Version in the webapp footer reports as v0.8.3-0-g61da2c0.

**Comment by anders0ndg** [comment]
The PiBakery image is working with all updates applied! Now on to SmartThings integration.

Thank you!!
Duwayn

**Comment by dcmeglio** [comment]
@anders0ndg  @bradpeczka could either of you explain how I install pibakery?

**Comment by f34rdotcom** [comment]
Don't need it. Only if you want to modify my recipe etc.

All info is on pibakery page but in short it's just a GUI app to build and
flash rasp erry pi images and customize and automate this process. You
could for example use it to modify my image to configure the wifi settings
for your location. Although with raspbian 9 that is as easy as putting a
specific text file on the uSD boot disk using notepad before you put it
back in you pi and boot.

**Comment by dcmeglio** [comment]
@f34rdotcom OK, If I don't need it what should I be doing to correct the 502?

**Comment by bradpeczka** [comment]
Download the latest Stretch build from the Alarm Decoder website - PIBAKERY-AD2PIBOOT-RASPBIAN-9-STRETCH-20181015.

Flash it to your SD card like any other Pi image, boot your Pi, then perform the Alarm Decoder setup and run the updates as normal.

**Comment by anders0ndg** [comment]
Sean,

After I enabled the SmartThings integration (API key and UPNP Push notifications) the keypad in the webapp stays at “Please wait Loading...” and the device handler does not update with current panel status. I can arm and disarm the panel from SmartThings, but devices/contacts do not show current status either. 

Thanks in advance,
Duwayn
918-671-5322

**Comment by melikyuksel** [comment]
The update killed my instances too. "Services might still be starting - please wait and try again. If this continues, please let us know on GitHub. You can also email us at Support." Shouldn't have updated while away from home, but didn't expect an update to brick my device...

**Comment by f34rdotcom** [comment]
@melikyuksel sadly the new code is not compatible with the old image. That image reached the end of support patches a while back so I made a new image for everyone to use. Everyone who wants to upgrade needs to backup your settings and update your flash to the new image then restore your settings. If you used the old SmartThings notification plugin you can remove it and use the new UPNP plugin as described on the SmartThings/Hubitat driver readme.md If you need help flashing feel free to contact me directly at ad2usb@support.nutech.com to schedule a time for me to help. It may be a good time to swap out the uSD card with a newer larger class 10 card. Keep your old card if you need your settings the are in /opt/alarmdecoder-webapp/instance/db.sqlite

AFAIK this is an LTS Raspbian image so it should like the last image maintain support for about 5 years.

Best
 Sean M

**Comment by melikyuksel** [comment]
@f34rdotcom Is there a way to revert to the old version in the meantime so that I can use the alarm system until I get to the site to re-image everything? I’m assuming simply git checking out old versions won’t work, and I’m not even sure which commit was the last one that worked.

**Comment by f34rdotcom** [comment]
Sure here is the command that "should" work. 
First ssh into the pi and stop gunicorn 
  'sudo service gunicorn stop' 
Then enter the web app folder.
  'cd /opt/alarmdecoder-webapp/'
Now checkout the last release.
 git checkout 0.8.2
Now start gunicorn again.
  'sudo service gunicorn start'

Let me know how this goes so others can use the same procedure.

**Comment by melikyuksel** [comment]
@f34rdotcom That didn’t work, still fails to start. Says “no module named ext.babel” in syslog before exiting.

**Comment by f34rdotcom** [comment]
What happens if you go back to 0.8.1?
Best
 SM

**Comment by melikyuksel** [comment]
I followed this saintly person’s instructions and got it working: http://www.alarmdecoder.com/forums/viewtopic.php?f=3&t=1101#p3919

Manually updating Flask using pip, and then editing line 33 of `/usr/local/lib/python2.7/dist-packages/flask_cache/jinja2ext.py` as he wrote seems to have worked in getting gunicorn to start properly and everything appears to work. What a mess!

**Comment by f34rdotcom** [comment]
So does that mean you can move to 0.8.3?

**Comment by melikyuksel** [comment]
Not sure, I checked out the master branch before making the changes. So 0.8.3 is probably fine after the changes.

**Comment by f34rdotcom** [comment]
Ok well its a fix so Thanks! and sorry for all this pain on this update. I think I will add to the update page a warning and a link to the config backup as well as instructions on how to manually change branches. It may get a bit wordy.

**Comment by f34rdotcom** [comment]
The warning on the update screen. Maybe future a way to upload settings before the wizard :(
